### ISSUE TITLE
Prometheus Metrics Implementation

### ISSUE THREAD

**Issue by prestwich** [celo-org/optics-monorepo#657]
Epic container for Agent Prometheus Metrics

Relevant Discussion: https://github.com/celo-org/optics-monorepo/discussions/645

**Comment by emberian** [comment]
>  metrics mod contains once_cell for a CoreMetrics struct

what do you mean by this? the global state lives under this namespace?

**Comment by emberian** [comment]
https://prometheus.io/docs/concepts/data_model/

**Comment by emberian** [comment]
wallet address doesn't make sense to me as a metric.

**Comment by prestwich** [comment]
> > metrics mod contains once_cell for a CoreMetrics struct
> 
> what do you mean by this? the global state lives under this namespace?

metrics that are common across all agents should be defined in a struct in `optics_base::metrics`. If we want to make these into global state, use `once_cell`, not `lazy_static` ( [see here for example](https://docs.rs/prometheus/0.12.0/prometheus/#static-metrics) )

**Comment by prestwich** [comment]
> wallet address doesn't make sense to me as a metric.

not really, but it's kinda useful to display them alongside the metrics

**Comment by emberian** [comment]
# metric naming

Using prometheus terminology, metrics have a standard format of `namespace_subsystem_name_units`. `optics` will be the `namespace`. Should the agents have their own metrics, making `subsystem` an agent name, or should the particular agent instead be a `label`? Labels are a way of slicing up datapoints recorded under a single metric along arbitrary dimensions, specified in the opentsdb format inside braces, such as `optics_updater_wallet_balance_total` vs `optics_wallet_balance_total{agent="updater"}`.

**Comment by emberian** [comment]
# datatype compatibility

prometheus only supports double and uint64, excluding in full generality wallet balance from being a metric datapoint. should we truncate to 64 bits? convert to float?
